src/libsync/clientsideencryption.cpp

```diff
@@ -0,0 +1,1168 @@
+    auto cnArray = _account->davUser().toLocal8Bit();
+    qCInfo(lcCse()) << "Getting the following array for the account Id" << cnArray;
+
+    auto certParams = std::map<const char *, const char*>{
+      {"C", "DE"},
+      {"ST", "Baden-Wuerttemberg"},
+      {"L", "Stuttgart"},
+      {"O","Nextcloud"},
+      {"CN", cnArray.constData()}
+    };
+
+    int             ret = 0;
+    int             nVersion = 1;
+
+    X509_REQ *x509_req = nullptr;
+    auto out = BIO_new(BIO_s_mem());
+    QByteArray output;
+    char data[80];
+    SignPublicKeyApiJob *job = nullptr;
+
+    // 2. set version of x509 req
+    x509_req = X509_REQ_new();
+    ret = X509_REQ_set_version(x509_req, nVersion);
+
+    // 3. set subject of x509 req
+    auto x509_name = X509_REQ_get_subject_name(x509_req);
+
+    using ucharp = const unsigned char *;
+    for(const auto& v : certParams) {
+        ret = X509_NAME_add_entry_by_txt(x509_name, v.first,  MBSTRING_ASC, (ucharp) v.second, -1, -1, 0);
+        if (ret != 1) {
+            qCInfo(lcCse()) << "Error Generating the Certificate while adding" << v.first << v.second;
+            goto free_all;
+        }
+    }
+
+    ret = X509_REQ_set_pubkey(x509_req, keyPair);
+    if (ret != 1){
+        qCInfo(lcCse()) << "Error setting the public key on the csr";
+        goto free_all;
+    }
+
+    ret = X509_REQ_sign(x509_req, keyPair, EVP_sha1());    // return x509_req->signature->length
+    if (ret <= 0){
+        qCInfo(lcCse()) << "Error setting the public key on the csr";
+        goto free_all;
+    }
+
+    ret = PEM_write_bio_X509_REQ(out, x509_req);
+    do {
+        ret = BIO_gets(out, data, 80);
+        output += data;
+        if (output.endsWith("-----END CERTIFICATE REQUEST-----")) {
+            output = output.trimmed();
+            break;
+        }
+    } while (ret > 0 );
+
+    qCInfo(lcCse()) << "Returning the certificate";
+    qCInfo(lcCse()) << output;
+
+    job = new SignPublicKeyApiJob(_account, baseUrl() + "public-key", this);
+    job->setCsr(output);
+
+    connect(job, &SignPublicKeyApiJob::jsonReceived, [this, keyPair](const QJsonDocument& json, int retCode) {
+        if (retCode == 200) {
+            auto caps = json.object().value("ocs").toObject().value("data").toObject().value("public-key").toString();
+            qCInfo(lcCse()) << "Public Key Returned" << caps;
+            QFile file(publicKeyPath(_account) + ".sign");
+            if (file.open(QIODevice::WriteOnly)) {
+                QTextStream s(&file);
+                s << caps;
+            }
+            file.close();
+            qCInfo(lcCse()) << "public key saved, Encrypting Private Key.";
+            encryptPrivateKey(keyPair);
+        }
+        qCInfo(lcCse()) << retCode;
+    });
+    job->start();
+
+free_all:
+    X509_REQ_free(x509_req);
+    BIO_free_all(out);
+}
+
+void ClientSideEncryption::setTokenForFolder(const QByteArray& folderId, const QByteArray& token)
+{
+	_folder2token[folderId] = token;
```

Comment:
I was thinking about this. *maybe* we should store this more permanent. Since if the client crashes, or shuts down, or whatever, and it is still locked we won't be able to unlock the folder ever again.